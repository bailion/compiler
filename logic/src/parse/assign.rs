use std::fmt::{self, Write};

use super::{expr::Expr, ident::Ident, utils::Parse};

#[derive(Debug, PartialEq, Eq)]
pub struct Assignment<'a> {
    pub to: Ident<'a>,
    pub value: Expr<'a>,
}

impl<'a> Parse<'a> for Assignment<'a> {
    fn parse(input: &mut super::utils::Input<'a>) -> Result<Self, super::utils::ParseError<'a>> {
        let to = Ident::parse(input)?;

        input.skip_whitespace()?;
        input.parse_token("=")?;
        input.skip_whitespace()?;
        let value = Expr::parse(input)?;

        Ok(Self { to, value })
    }
}

impl fmt::Display for Assignment<'_> {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        self.to.fmt(f)?;
        f.write_char('=')?;
        self.value.fmt(f)
    }
}

#[cfg(test)]
mod test_simple_assignment {
    use proptest::prelude::*;

    use crate::parse::{
        assign::Assignment,
        utils::{Input, Parse},
    };
    #[cfg(feature = "_proptest")]
    proptest! {
    #[test]
        fn test(input in "[a-zA-Z0-9]+[ ]*=[ ]*[a-zA-Z0-9]+") {
            test_inner(&input);
        }
    }

    fn test_inner(input: &str) {
        let mut input = Input::new(input);
        Assignment::parse(&mut input).expect("faild to parse the valid assignment");
        assert!(input.is_empty())
    }

    #[test]
    fn regressions() {
        test_inner("A=A");
    }
}
