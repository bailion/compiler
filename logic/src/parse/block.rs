use std::fmt;

use super::{utils::Parse, Ast};

#[derive(Debug, PartialEq, Eq)]
pub struct Block<'a> {
    inner: Ast<'a>,
    indent: usize,
}

impl fmt::Display for Block<'_> {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        self.inner.fmt(f)
    }
}

impl<'a> Parse<'a> for Block<'a> {
    fn parse(input: &mut super::utils::Input<'a>) -> Result<Self, super::utils::ParseError<'a>> {
        input.increment_indent(2);
        let res = Ok(Self {
            inner: Ast::parse(input)?,
            indent: input.indent,
        });
        input.decrement_indent(2);
        res
    }
}
