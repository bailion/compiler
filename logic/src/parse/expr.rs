use std::fmt::{self, Display, Write};

use super::{
    ident::Ident,
    lit::Literal,
    utils::{Input, Parse, ParseError},
};

#[derive(Debug, PartialEq, Eq)]
pub enum Expr<'a> {
    Ident(Ident<'a>),
    Literal(Literal<'a>),
    BinOp(BinOp, Box<Expr<'a>>, Box<Expr<'a>>),
    UnOp(UnOp, Box<Expr<'a>>),
}

impl<'a> Parse<'a> for Expr<'a> {
    fn parse(input: &mut super::utils::Input<'a>) -> Result<Self, super::utils::ParseError<'a>> {
        Self::parse_bp(input, 0).and_then(|op| op.ok_or(ParseError::__NonExhaustive))
    }
}

impl fmt::Display for Expr<'_> {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        match self {
            Expr::Ident(ident) => ident.fmt(f),
            Expr::BinOp(operator, left, right) => {
                left.fmt(f)?;
                f.write_char(' ')?;
                operator.fmt(f)?;
                f.write_char(' ')?;
                right.fmt(f)
            }
            Expr::UnOp(op, left) => {
                op.fmt(f)?;
                f.write_char(' ')?;
                left.fmt(f)
            }
            Expr::Literal(lit) => lit.fmt(f),
        }
    }
}

impl<'a> Expr<'a> {
    pub(crate) fn parse_bp(
        input: &mut Input<'a>,
        min_bp: u8,
    ) -> Result<Option<Self>, ParseError<'a>> {
        Self::parse_bp_stop_if(input, min_bp, |_| false)
    }

    pub(crate) fn parse_bp_stop_if(
        input: &mut Input<'a>,
        min_bp: u8,
        stop_if: impl Fn(&str) -> bool,
    ) -> Result<Option<Self>, ParseError<'a>> {
        input.skip_whitespace()?;
        let mut lhs = {
            if let Ok(_) = Ident::parse(&mut input.clone()) {
                let ident = Ident::parse(input)?;

                Some(Self::Ident(ident))
            } else if let Ok(_) = Literal::parse(&mut input.clone()) {
                let lit = Literal::parse(input)?;
                dbg!(&lit);
                dbg!(&input);

                Some(Self::Literal(lit))
            } else {
                None
            }
        };

        loop {
            input.skip_whitespace()?;

            if input.is_empty() || input.starts_with('\n') || (stop_if)(&*input) {
                break;
            }

            let op = match Op::parse(&mut input.clone(), lhs.is_none()) {
                Ok(op) => op,
                Err(err) => match err {
                    ParseError::UnexpectedEndOfInput => break,
                    e => {
                        return Err(e);
                    }
                },
            };

            let (left_bp, right_bp) = op.bp();
            if left_bp < min_bp {
                break;
            }

            Op::parse(input, false)?;

            let rhs = Self::parse_bp(input, right_bp)?;

            if op.is_bin_op() {
                match (lhs, rhs) {
                    (Some(left), Some(right)) => {
                        lhs = Some(Self::BinOp(
                            op.try_into_bin_op().unwrap(),
                            Box::new(left),
                            Box::new(right),
                        ));
                    }
                    _ => return Err(ParseError::__NonExhaustive),
                }
            } else {
                if let Some(rhs) = rhs {
                    lhs = Some(Self::UnOp(op.try_into_un_op().unwrap(), Box::new(rhs)))
                } else {
                    return Err(ParseError::__NonExhaustive);
                }
            }
        }

        Ok(lhs)
    }
}

#[derive(Copy, Clone, Debug)]
pub enum Op {
    BinOp(BinOp),
    UnOp(UnOp),
}

impl Op {
    fn parse<'a>(input: &mut Input<'a>, prefix: bool) -> Result<Self, ParseError<'a>> {
        input.skip_whitespace()?;
        Ok(match input.advance_one()? {
            "+" if !prefix => Op::BinOp(BinOp::Add),
            "-" if !prefix => Op::BinOp(BinOp::Subtract),
            "/" => Op::BinOp(BinOp::Divide),
            "*" => Op::BinOp(BinOp::Multiply),
            "+" if prefix => Op::UnOp(UnOp::Positive),
            "-" if prefix => Op::UnOp(UnOp::Negative),
            "=" => Op::BinOp(BinOp::SetEquals),
            token => {
                return Err(ParseError::UnexpectedToken {
                    token,
                    explanation: format!(
                        "Expected an operator here. Instead we found `{}`.",
                        token
                    ),
                })
            }
        })
    }

    fn bp(&self) -> (u8, u8) {
        match self {
            Op::BinOp(op) => op.bp(),
            Op::UnOp(op) => op.bp(),
        }
    }

    pub fn try_into_bin_op(self) -> Result<BinOp, Self> {
        match self {
            Self::BinOp(v) => Ok(v),
            Op::UnOp(_) => Err(self),
        }
    }

    pub fn try_into_un_op(self) -> Result<UnOp, Self> {
        if let Self::UnOp(v) = self {
            Ok(v)
        } else {
            Err(self)
        }
    }

    /// Returns `true` if the op is [`BinOp`].
    pub fn is_bin_op(&self) -> bool {
        matches!(self, Self::BinOp(..))
    }
}

#[derive(Debug, PartialEq, Eq, Copy, Clone)]
pub enum BinOp {
    Add,
    Subtract,
    Divide,
    Multiply,
    SetEquals,
}

impl BinOp {
    fn bp(&self) -> (u8, u8) {
        match self {
            BinOp::Add | BinOp::Subtract => (5, 6),
            BinOp::Divide | BinOp::Multiply => (7, 8),
            BinOp::SetEquals => (2, 1),
        }
    }
}

impl fmt::Display for BinOp {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        f.write_str(match self {
            BinOp::Add => "+",
            BinOp::Subtract => "-",
            BinOp::Divide => "/",
            BinOp::Multiply => "*",
            BinOp::SetEquals => "=",
        })
    }
}

#[derive(Debug, Copy, Clone, PartialEq, Eq)]
pub enum UnOp {
    Positive,
    Negative,
}

impl UnOp {
    fn bp(&self) -> (u8, u8) {
        match self {
            UnOp::Positive | UnOp::Negative => (99, 9),
        }
    }
}

impl Display for UnOp {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        f.write_str(match self {
            UnOp::Positive => "+",
            UnOp::Negative => "-",
        })
    }
}
