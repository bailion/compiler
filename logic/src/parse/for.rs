use core::fmt;
use std::{fmt::Write, marker::PhantomData};

use super::{
    block::Block,
    expr::Expr,
    ident::Ident,
    utils::{write_indentation, Parse, ParseError},
};

#[derive(Debug, PartialEq, Eq)]
pub struct ForLoop<'a, IDENT = Ident<'a>, EXPR = Expr<'a, IDENT>> {
    pub(crate) var: IDENT,
    pub(crate) between: Between<'a, IDENT, EXPR>,
    pub(crate) block: Block<'a, IDENT, EXPR>,
    pub(crate) indent: usize,
}

impl<'a> Parse<'a> for ForLoop<'a> {
    fn parse(input: &mut super::utils::Input<'a>) -> Result<Self, super::utils::ParseError<'a>> {
        input.parse_token("for")?;
        input.skip_whitespace()?;

        let var = Ident::parse(input)?;
        input.skip_whitespace()?;
        input.parse_token("=")?;
        input.skip_whitespace()?;
        let between = Between::parse(input)?;

        input.advance_whitespace_and_new_line()?;

        let block = Block::parse(input)?;

        input.parse_token("\n")?;

        input.advance_indent()?;
        input.parse_token("next")?;
        input.skip_whitespace()?;

        let ident = Ident::parse(input)?;

        if *ident != *var {
            return Err(ParseError::UnexpectedToken {
                token: *ident,
                explanation: format!(
                    "Expected the identifier `{}` here, but instead found `{}`.",
                    *var, *ident
                ),
            });
        }

        input.skip_whitespace()?;

        if !input.is_empty() && input.chars().next() != Some('\n') {
            return Err(ParseError::UnexpectedToken {
                token: input.peek_n(1).unwrap_or(""),
                explanation: "Expected a new line here.".to_string(),
            });
        }

        Ok(Self {
            var,
            between,
            block,
            indent: input.indent,
        })
    }
}

impl fmt::Display for ForLoop<'_> {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write_indentation(self.indent, f)?;
        f.write_str("for ")?;
        self.var.fmt(f)?;
        f.write_str(" = ")?;
        self.between.start.fmt(f)?;
        f.write_str(" to ")?;
        self.between.stop.fmt(f)?;
        if let Some(ref step) = self.between.step {
            f.write_str(" step ")?;
            step.fmt(f)?;
        }
        f.write_char('\n')?;
        self.block.fmt(f)?;
        f.write_char('\n')?;
        write_indentation(self.indent, f)?;
        f.write_str("next ")?;
        self.var.fmt(f)?;
        f.write_char('\n')
    }
}

#[derive(Debug, PartialEq, Eq)]
pub struct Between<'a, IDENT = Ident<'a>, EXPR = Expr<'a>> {
    pub(crate) start: EXPR,
    pub(crate) stop: EXPR,
    pub(crate) step: Option<EXPR>,
    pub(crate) _i: PhantomData<&'a IDENT>,
}

impl<'a> Parse<'a> for Between<'a> {
    fn parse(input: &mut super::utils::Input<'a>) -> Result<Self, super::utils::ParseError<'a>> {
        let start = Expr::parse_bp_stop_if(input, 0, |input| input.starts_with("to"))?
            .ok_or(ParseError::__NonExhaustive)?;

        input.skip_whitespace()?;

        input.parse_token("to")?;

        input.skip_whitespace()?;

        let stop = Expr::parse_bp_stop_if(input, 0, |input| input.starts_with("step"))?
            .ok_or(ParseError::__NonExhaustive)?;

        let step = if input.clone().parse_token("step").is_ok() {
            input.parse_token("step")?;
            Some(
                Expr::parse_bp_stop_if(input, 0, |input| input.starts_with('\n'))?
                    .ok_or(ParseError::__NonExhaustive)?,
            )
        } else {
            None
        };

        Ok(Self {
            start,
            stop,
            step,
            _i: PhantomData,
        })
    }
}
