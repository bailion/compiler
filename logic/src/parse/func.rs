use std::{
    fmt::{self, Write},
    marker::PhantomData,
};

use super::{
    block::Block,
    expr::Expr,
    ident::Ident,
    utils::{write_indentation, Parse},
};

#[derive(Debug, Eq, PartialEq)]
pub struct Func<'a, IDENT = Ident<'a>, EXPR = Expr<'a>> {
    pub(crate) name: IDENT,
    pub(crate) parameters: Vec<IDENT>,
    pub(crate) block: Block<'a, IDENT, EXPR>,
    pub(crate) indent: usize,
}

impl<'a> Parse<'a> for Func<'a> {
    fn parse(input: &mut super::utils::Input<'a>) -> Result<Self, super::utils::ParseError<'a>> {
        input.advance_indent()?;
        input.parse_token("function")?;
        input.skip_whitespace()?;
        let name = Ident::parse(input)?;
        input.skip_whitespace()?;
        input.parse_token("(")?;

        let parameters = input.delimited_list(Ident::parse, ')', ",")?;

        input.parse_token(")")?;

        input.skip_whitespace()?;
        input.parse_token("\n")?;
        let block = Block::parse(input)?;
        input.advance_indent()?;
        input.parse_token("endfunction")?;
        input.skip_whitespace()?;
        Ok(Self {
            name,
            parameters,
            block,
            indent: input.indent,
        })
    }
}

impl fmt::Display for Func<'_> {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write_indentation(self.indent, f)?;
        f.write_str("function ")?;
        self.name.fmt(f)?;
        f.write_str(" (")?;
        for param in &self.parameters {
            param.fmt(f)?;
            f.write_char(',')?;
        }
        f.write_str(")\n")?;
        self.block.fmt(f)?;
        f.write_char('\n')?;
        write_indentation(self.indent, f)?;
        f.write_str("endfunction")?;
        f.write_char('\n')
    }
}

#[derive(Debug, Eq, PartialEq)]
pub struct Return<'a, EXPR = Expr<'a>> {
    pub(crate) expr: EXPR,
    pub(crate) indent: usize,
    pub(crate) _a: PhantomData<&'a ()>,
}

impl<'a> Parse<'a> for Return<'a> {
    fn parse(input: &mut super::utils::Input<'a>) -> Result<Self, super::utils::ParseError<'a>> {
        input.parse_token("return ")?;

        Ok(Self {
            expr: Expr::parse(input)?,
            indent: input.indent,
            _a: PhantomData,
        })
    }
}

impl fmt::Display for Return<'_> {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        f.write_str("return ")?;
        self.expr.fmt(f)
    }
}
