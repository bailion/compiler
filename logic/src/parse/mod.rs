//! The parser for programs. It's a recursive descent parser which I try to make emit friendly and
//! helpful error messages.
//!
//! There is something approaching a formal grammar in [`crate::parse::fuzz`].

use self::{
    expr::Expr,
    func::{Func, Return},
    r#for::ForLoop,
    r#if::If,
    r#while::While,
    record::Record,
    table::{ItemRef, ParseContext},
    utils::{Input, Parse, ParseError},
};

pub mod r#block;
pub mod expr;
pub mod r#for;
pub mod func;
pub mod ident;
pub mod r#if;
pub mod lit;
pub mod record;
pub mod table;
pub mod utils;
pub mod r#while;

#[cfg(all(test, fuzzing))]
mod fuzz;

#[cfg(test)]
mod test;
#[cfg(test)]
mod test_scopes;
#[cfg(test)]
/// Tests to check that the compiler produces well-formed and comprehensible
/// error messages.
pub mod ui;

pub use table::parse;

/// Parses a list of statements, returning a list of [`ItemRef`]s.
fn parse_statements<'i>(
    input: &mut Input<'i>,
    ctx: &mut ParseContext<'i>,
) -> Result<Vec<ItemRef>, ParseError> {
    let mut nodes = vec![];
    loop {
        // remove any lines which only consist of whitespace characters
        loop {
            let mut tmp = *input;
            tmp.skip_whitespace()?;
            if tmp.starts_with('\n') {
                input.skip_whitespace()?;
                input.parse_token("\n")?;
            } else {
                break;
            }
        }

        if input.is_empty()
            // note: we rely here on the fact that `&&` is short-circuiting (i.e. if the first
            // operand is false, then the second operand should not be evaluated)
            || (input.indent >= 2 && input.count_indent()? == input.indent - 2)
            || input.chars().all(|char| char.is_whitespace())
        {
            return Ok(nodes);
        } else {
            input.advance_indent()?;
            // comments
            if input.starts_with(";;") {
                input.eat_until_or_end(|c| c == '\n')?;
            } else {
                nodes.push(Node::parse(input, ctx)?);
            }
        }
    }
}

#[cfg(test)]
#[test]
fn empty_statements() {
    assert_eq!(
        parse_statements(&mut Input::new(""), &mut ParseContext::new()),
        Ok(Vec::new())
    );
}

#[derive(Debug, PartialEq, Eq)]
pub struct Node;

impl<'i> Parse<'i> for Node {
    type Output = ItemRef;
    type Context = ParseContext<'i>;

    fn parse(
        input: &mut utils::Input<'i>,
        ctx: &mut ParseContext<'i>,
    ) -> Result<ItemRef, ParseError> {
        if input.starts_with("for ") {
            ForLoop::parse(input, ctx).map(From::from)
        } else if input.starts_with("return ") {
            Return::parse(input, ctx).map(From::from)
        } else if input.starts_with("if ") {
            If::parse(input, ctx).map(From::from)
        } else if input.starts_with("while ") {
            While::parse(input, ctx).map(From::from)
        } else if input.starts_with("function ") {
            Func::parse(input, ctx).map(From::from)
        } else if input.starts_with("record") {
            Record::parse(input, ctx).map(From::from)
        } else {
            Expr::parse(input, ctx).map(From::from)
        }
    }
}
