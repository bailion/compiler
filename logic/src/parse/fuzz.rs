use fuzzcheck::mutators::grammar::*;

use super::parse;

fn run_test(input: &str) -> bool {
    if let Ok(ast) = parse(input) {
        let out = ast.to_string();
        let reconstructed_ast = if let Ok(ast) = parse(&out) {
            ast
        } else {
            return false;
        };
        ast == reconstructed_ast
    } else {
        false
    }
}

#[test]
fn fuzz_parser() {
    let ident = regex("[a-hA-Z][a-h0-9A-Z_]+");

    let string = concatenation([literal('"'), ident.clone(), literal('"')]);

    let number = regex("[0-9]+");

    let float = concatenation([
        number.clone(),
        literal('.'),
        number.clone(),
        literal('e'),
        number.clone(),
    ]);

    let expression = recursive(|e| {
        alternation([
            ident.clone(),
            string.clone(),
            number.clone(),
            float.clone(),
            // function calls
            concatenation([
                ident.clone(),
                literal('('),
                repetition(concatenation([recurse(e), literal(',')]), 1..),
                literal(')'),
            ]),
            concatenation([literal('('), recurse(e), literal(')')]),
            concatenation([
                literal('('),
                recurse(e),
                regex("[-+*/]"),
                recurse(e),
                literal(')'),
            ]),
            concatenation([regex("[-+]"), literal('('), recurse(e), literal(')')]),
        ])
    });

    let statement = concatenation([
        alternation([
            expression.clone(),
            concatenation([ident.clone(), literal('='), expression.clone()]),
            concatenation([regex("return "), expression.clone()]),
        ]),
        literal('\n'),
    ]);

    let for_loop_ident = literal('i');

    let indented_statements = repetition(concatenation([regex("[\n]+  "), statement.clone()]), 1..);

    let for_loop = concatenation([
        regex("for "),
        for_loop_ident.clone(),
        regex("= "),
        expression.clone(),
        regex(" to "),
        expression.clone(),
        regex(" step "),
        expression.clone(),
        literal('\n'),
        indented_statements.clone(),
        regex("next "),
        for_loop_ident,
        literal('\n'),
    ]);

    let while_loop = concatenation([
        regex("while "),
        expression.clone(),
        literal('\n'),
        indented_statements.clone(),
        regex("endwhile\n"),
    ]);

    let function = concatenation([
        regex("function "),
        ident.clone(),
        literal('('),
        repetition(concatenation([ident.clone(), literal(',')]), 1..),
        literal(')'),
        literal('\n'),
        indented_statements.clone(),
        regex("endfunction\n"),
    ]);

    let if_statement = concatenation([
        regex("if "),
        expression.clone(),
        regex(" then\n"),
        indented_statements.clone(),
        repetition(
            concatenation([
                regex("elseif "),
                expression.clone(),
                regex(" then\n"),
                indented_statements.clone(),
            ]),
            0..,
        ),
        repetition(
            concatenation([regex("else\n"), indented_statements.clone()]),
            0..=1,
        ),
        regex("endif\n"),
    ]);

    let block = alternation([
        for_loop.clone(),
        while_loop.clone(),
        function.clone(),
        if_statement,
    ]);

    let ast = repetition(alternation([block, statement]), 0..);

    let ast_mutator = grammar_based_ast_mutator(ast);

    let result = fuzzcheck::fuzz_test(|ast: &AST| run_test(&ast.to_string()))
        .mutator(ast_mutator)
        .serde_serializer()
        .default_sensor_and_pool()
        .arguments_from_cargo_fuzzcheck()
        .launch();
    assert!(!result.found_test_failure);
}
