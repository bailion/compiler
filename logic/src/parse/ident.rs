use std::{fmt::{self, Write}, hash};

use crate::diagnostics::span::Span;

use super::utils::{Parse, ParseError};

pub const KEYWORDS: &[&'static str] = &[&"for", &"if", &"while", &"function", &"return"];

#[derive(Debug, Copy, Clone, PartialEq, Eq)]
pub struct Ident<'a> {
    inner: &'a str,
    span: Span,
}

impl<'a> hash::Hash for Ident<'a> {
    fn hash<H: hash::Hasher>(&self, state: &mut H) {
        self.inner.hash(state)
    }
}

impl<'a> std::ops::Deref for Ident<'a> {
    type Target = &'a str;

    fn deref(&self) -> &Self::Target {
        &self.inner
    }
}

impl<'a> Parse<'a> for Ident<'a> {
    fn parse(input: &mut super::utils::Input<'a>) -> Result<Self, super::utils::ParseError<'a>> {
        input.skip_whitespace()?;
        let rec = input.start_recording();
        input
            .eat_until_or_end(|char| !char.is_alphanumeric() && char != '_')
            .and_then(|inner| {
                if inner.is_empty() {
                    Err(ParseError::UnexpectedEndOfInput)
                } else if !inner.chars().next().unwrap().is_alphabetic() {
                    Err(ParseError::__NonExhaustive)
                } else if KEYWORDS.contains(&inner) {
                    Err(ParseError::UnexpectedToken {
                        token: inner,
                        explanation: format!(
                            "Expected an identifier here, but `{}` is a keyword.",
                            inner
                        ),
                    })
                } else {
                    Ok(Self {
                        inner,
                        span: rec.finish_recording(input),
                    })
                }
            })
    }
}

impl fmt::Display for Ident<'_> {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        self.inner.fmt(f)?;
        f.write_char(' ')
    }
}

#[cfg(test)]
mod test_parse_valid_idents {
    #[cfg(feature = "_proptest")]
    use proptest::prelude::*;

    use crate::parse::{
        ident::Ident,
        utils::{Input, Parse},
    };

    #[cfg(feature = "_proptest")]
    proptest! {
        #[test]
        fn test(input in "[a-zA-Z0-9]+") {
            test_inner(&input)
        }
    }

    fn test_inner(string: &str) {
        let mut input = Input::new(string);
        let ident = Ident::parse(&mut input).expect("failed to parse");
        assert!(input.is_empty());
        assert_eq!(string, ident.inner);
    }

    #[test]
    fn regressions() {
        test_inner("x");
        test_inner("function_call");
    }

    #[test]
    fn test_invalid_idents_do_not_parse() {
        Ident::parse(&mut Input::new("////")).expect_err("this identifier should not be valid");
    }
}
