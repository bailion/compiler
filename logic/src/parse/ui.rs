use codespan_reporting::{
    files::SimpleFiles,
    term::{self, termcolor::Buffer, Config},
};

use super::parse;

fn setup_file(file: String) -> (usize, SimpleFiles<String, String>) {
    let mut files = SimpleFiles::new();
    let id = files.add("file".to_string(), file);
    (id, files)
}

fn ui_test(string: String) {
    let err = parse(&string).unwrap_err();
    let (file_id, files) = setup_file(string);
    let report = err.report(file_id);

    let mut writer = Buffer::no_color();

    let config = Config::default();

    term::emit(&mut writer, &config, &files, &report).unwrap();

    let vec = writer.into_inner();
    let string = String::from_utf8(vec).unwrap();

    insta::assert_snapshot!(string);
}

#[test]
fn test_invalid_operator() {
    ui_test("(a b c)".to_string())
}

#[test]
fn test_invalid_brackets_1() {
    ui_test("( ( ) a".to_string())
}

#[test]
fn test_invalid_brackets2() {
    ui_test("f(x + a * z (( zz".to_string())
}
