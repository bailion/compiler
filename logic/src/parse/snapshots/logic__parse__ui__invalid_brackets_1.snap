---
source: logic/src/parse/ui.rs
expression: string

---
error: Your program contains a syntax error!
  ┌─ file:1:7
  │
1 │ ( ( ) a
  │       ^ Expected an operator here. Instead we found `a`.


